module/zfs/zfs_znode.c

```diff
@@ -1049,6 +1075,11 @@ zfs_xvattr_set(znode_t *zp, xvattr_t *xvap, dmu_tx_t *tx)
 		    zp->z_pflags, tx);
 		XVA_SET_RTN(xvap, XAT_SPARSE);
 	}
+	if (XVA_ISSET_REQ(xvap, XAT_PROJINHERIT)) {
+		ZFS_ATTR_SET(zp, ZFS_PROJINHERIT, xoap->xoa_projinherit,
+		    zp->z_pflags, tx);
```

Comment:
Do I understand correctly that this is setting or clearing the PROJINHERIT flag on a znode at the user's request?  How do we enforce the invariant that if PROJINHERIT is set, all the descendents have the same PROJID, and all descendent directories have PROJINHERIT also set?  It seems like the PROJINHERIT flag can only be changed efficiently (in O(1) time) if the directory is empty.  Even then we would have to check if the parent has PROJINHERIT set (and not allow clearing PROJINHERIT if the parent has it set).

Reply:
After reading all the various documentation I could find about project quotas I was left with the impression that the original design intention was to enforce tree quotas.  For example, the `chattr(1)` man page clearly states:

> A directory with the 'P' attribute  set  will  **enforce**  a  hierarchical structure  for  project id's.

It was the need for the `xfs_quota project -d` depth option and ability for a privileged user to override a PROJID which complicated the design.  Unfortunately, it wasn't clear to me from any of my reading what use case persuaded the developers to add this functionality.  Maybe @adilger has some insight in to why it was added.

Assuming that we can drop that functionality (and I think we should barring a compelling use case), I like @ahrens proposal for how we can enforce this on the kernel side.  Using the `zfs project` command to manage the trees also makes good sense but there are going to be a few complications we'll need to decide how to handle.

For example, `zfs project -s` could fail in a tree where new files are being added unless we allow setting PROJINHERIT on directories to be retried.  We should also consider how to handle a `zfs project -s` failure which leaves a tree partially set up.  This would be a legitimate case where `zfs project -c` would report files which do not have the PROJID for the rest of the tree.

As for the `/etc/projects` file my understanding is it's shorthand so you don't need to always use the `-p` option to specify the full path.  How helpful that is in practice I'm not really sure, I wouldn't object to skipping this.  The `/etc/projid` file on the other hand I definitely thing we should support if it exists.

Reply:
> unless we allow setting PROJINHERIT on directories to be retried

I think you mean, "unless userland will retry setting ..." since of course it would be allowed.  FWIW, I don't think that similar userland commands retry, they simply fail if there are concurrent changes to the directory tree being modified.  E.g. "rm -r" or "chmod -r".

> This would be a legitimate case where zfs project -c would report files which do not have the PROJID for the rest of the tree

I think that `zfs project -c` would still be O(1), it just checks if PROJINHERIT is set on the specified directory.  I guess if it's not set, we could have it traverse down and see if it's set anywhere below?  But that doesn't seem necessary.

> The /etc/projid file on the other hand I definitely thing we should support if it exists.

I agree